src/meta_memcache/base/base_cache_pool.py

```diff
@@ -0,0 +1,210 @@
+        a fallback pool, etc.
+        """
+        try:
+            with self._get_pool(key).get_connection() as c:
+                return self._exec_on_connection(
+                    c,
+                    command=command,
+                    key=key,
+                    value=value,
+                    flags=flags,
+                    int_flags=int_flags,
+                    token_flags=token_flags,
+                )
+        except MemcacheServerError:
+            if self._write_failure_tracker:
+                if command in (MetaCommand.MD, MetaCommand.MS):
+                    self._write_failure_tracker.add_key(key)
+            raise
+
+    def _exec_on_connection(
+        self,
+        c: MemcacheSocket,
+        command: MetaCommand,
+        key: Key,
+        value: Optional[bytes] = None,
+        flags: Optional[Set[Flag]] = None,
+        int_flags: Optional[Dict[IntFlag, int]] = None,
+        token_flags: Optional[Dict[TokenFlag, bytes]] = None,
+    ) -> Union[Miss, Value, Success, NotStored, Conflict]:
+        """
+        Execute command on a connection and read response back
+        """
+        cmd = self._build_cmd(
+            command,
+            key,
+            size=len(value) if value is not None else None,
+            flags=flags,
+            int_flags=int_flags,
+            token_flags=token_flags,
+        )
+        if value:
+            c.sendall(cmd + value + ENDL)
+        else:
+            c.sendall(cmd)
+        if flags and Flag.NOREPLY in flags:
+            return Success(flags=set([Flag.NOREPLY]))
+        result = c.get_response()
+        if isinstance(result, Value):
+            # TODO: Confirm this works for empty values!
+            data = c.get_value(result.size)
+            if result.size > 0:
+                encoding_id = result.int_flags.get(IntFlag.CLIENT_FLAG, 0)
+                result.value = self._serializer.unserialize(data, encoding_id)
+        return result
+
+    def mg(
+        self,
+        key: Key,
+        flags: Optional[Set[Flag]] = None,
+        int_flags: Optional[Dict[IntFlag, int]] = None,
+        token_flags: Optional[Dict[TokenFlag, bytes]] = None,
+    ) -> Union[Miss, Value, Success]:
+        result = self._exec(
+            command=MetaCommand.MG,
+            key=key,
+            flags=flags,
+            int_flags=int_flags,
+            token_flags=token_flags,
+        )
+        if not isinstance(result, (Miss, Value, Success)):
+            raise MemcacheError(
+                f"Unexpected response for Meta Delete command: {result}"
+            )
+        return result
+
+    def ms(
+        self,
+        key: Key,
+        value: Any,  # pyre-ignore[2]
+        ttl: int,
+        flags: Optional[Set[Flag]] = None,
+        int_flags: Optional[Dict[IntFlag, int]] = None,
+        token_flags: Optional[Dict[TokenFlag, bytes]] = None,
+    ) -> Union[Success, NotStored, Conflict, Miss]:
```

Comment:
Why don't return just one object with all the data inside like a HTTP response? That why you don't have to do `isinstance` to return a bool and you can return more rich responses if needed

Reply:
Some responses like miss don't have flags. Only value has size...

In general, it is better for string typing to use this mixed response. This forces callers to do isinstance(Value) before asking `.size` and saves a lot of runtime issues,

Reply:
The problem with that is that it is not very type safe.

Returning a `Union`, you force the user of this api to do type checks before using a value, forcing correct error handling.

```
if isinstance(result, Value):
    result.value  # value must exist!!! This is totally type-safe
```

If you really prefer not to have a `Union` I can definitely create an inheritance tree from a common response. But some methods will still use Union as a `get` should not return `stored` for example.

This is also the low-level api, which gives you the raw power of the meta commands and all the data in the responses. And in this layer I still think it is best to use unions for the possible responses. Most users will use the get/set/delete methods that have a much simpler api.

Reply:
> The problem with that is that it is not very type safe.

Why?

> If you really prefer not to have a Union I can definitely create an inheritance tree from a common response. But some methods will still use Union as a get should not return stored for example.

I'm not very opinionated, I'm trying to understand why is needed because for me returning different types for the same function is a bad smell and maybe we can have a better approach that is less error prone that having to use `isinstance` when we use that function

Reply:
Having multiple return types is not an anti pattern. Using unions and isinstance is good over a return object that has many optional values, with combinations of which don't make sense, and cause less strict handling on the caller.

In fact, `Optional[Foo]` is a type of union: `Union[None, Foo]`. When you return `Optional[Foo]` you force users to check for noneness before they can use the `Foo` properties (that is, if they use static type checks). This is very powerful to enforce correct error handling. 

Having `Union[Miss,Success,Value]` forces the caller to do the right handling. Only Value has data inside. With this you can potentially even store a `None` as a serialized value. You can ask callers to check "response type" and use the data only if response type is "value". But you see the problem here, it is not type safe, you can't be sure if users are really checking the reponse type or just using the optional data property directly. With a union, if you try to access `.data` on `Union[Miss, Value]` the type checker will yell at you saying that `data` is not a property of `Miss`.

Furthermore, I was even thinking in creating more response types for when Flag.WIN is returned, again, to make sure wins are handled explicitly, not mishandled as a Value. The win could for example not have data, as you want implementers to threat this as a miss. I didn't do it to mimic more closely the memcached api, and avoid confusion. It is after all an internal layer and I expect users of meta commands to be well aware of the quirks, but like I said, I think Unions on a statically type check python provide huge wins in robustness.